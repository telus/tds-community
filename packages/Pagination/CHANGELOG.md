# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.1.0](https://github.com/telus/tds-community/compare/@tds/community-pagination@2.0.4...@tds/community-pagination@2.1.0) (2021-09-21)


### Features

* **community-pagination:** react 17 and styled components 5 ([2ed4afb](https://github.com/telus/tds-community/commit/2ed4afb26f3208ad5b84fe7cc120588eea5e1ec5))





## [2.0.4](https://github.com/telus/tds-community/compare/@tds/community-pagination@2.0.3...@tds/community-pagination@2.0.4) (2020-09-09)


### Bug Fixes

* **community-pagination:** update util-helpers version ([5608ead](https://github.com/telus/tds-community/commit/5608ead0196621e3114300129c769721cdcbcab6))





## [2.0.3](https://github.com/telus/tds-community/compare/@tds/community-pagination@2.0.2...@tds/community-pagination@2.0.3) (2020-03-20)

**Note:** Version bump only for package @tds/community-pagination





## [2.0.2](https://github.com/telus/tds-community/compare/@tds/community-pagination@2.0.1...@tds/community-pagination@2.0.2) (2020-03-19)

**Note:** Version bump only for package @tds/community-pagination





## [2.0.1](https://github.com/telus/tds-community/compare/@tds/community-pagination@2.0.0...@tds/community-pagination@2.0.1) (2019-11-22)

**Note:** Version bump only for package @tds/community-pagination





# [2.0.0](https://github.com/telus/tds-community/compare/@tds/community-pagination@1.0.3...@tds/community-pagination@2.0.0) (2019-10-10)


### Bug Fixes

* **community-pagination:** design changes ([c648bc3](https://github.com/telus/tds-community/commit/c648bc3))
* **packages:** adding styled-components ([cf613ba](https://github.com/telus/tds-community/commit/cf613ba))


### Features

* **community-pagination:** design changes ([4d0d662](https://github.com/telus/tds-community/commit/4d0d662))


### BREAKING CHANGES

* **packages:** add styled-components to peerDependencies of Pagination
* **community-pagination:** Default visual designs have changed





## [1.0.3](https://github.com/telus/tds-community/compare/@tds/community-pagination@1.0.2...@tds/community-pagination@1.0.3) (2019-09-17)

**Note:** Version bump only for package @tds/community-pagination





## [1.0.2](https://github.com/telus/tds-community/compare/@tds/community-pagination@1.0.1...@tds/community-pagination@1.0.2) (2019-06-13)

**Note:** Version bump only for package @tds/community-pagination





## [1.0.1](https://github.com/telus/tds-community/compare/@tds/community-pagination@1.0.0...@tds/community-pagination@1.0.1) (2019-05-17)

**Note:** Version bump only for package @tds/community-pagination





# 1.0.0 (2019-04-18)


### Bug Fixes

* **community-pagination:** adjustment to elipsis ([acc713d](https://github.com/telus/tds-community/commit/acc713d))
* **community-pagination:** fixing spacing ([44dcef2](https://github.com/telus/tds-community/commit/44dcef2))
* **community-pagination:** safari bug fix ([5cb9ae7](https://github.com/telus/tds-community/commit/5cb9ae7))


### Features

* **community-pagination:** adding french accessibility text ([c578527](https://github.com/telus/tds-community/commit/c578527))
* **community-pagination:** adding unique keys to pagination ([6eb2a96](https://github.com/telus/tds-community/commit/6eb2a96))
* **community-pagination:** centering component ([1032622](https://github.com/telus/tds-community/commit/1032622))
* **community-pagination:** creating pagination component ([9ffeae9](https://github.com/telus/tds-community/commit/9ffeae9))
* **community-pagination:** french support,changes to styles,adding a11y ([6080093](https://github.com/telus/tds-community/commit/6080093))
* **community-pagination:** including tests and additional files ([5d3c382](https://github.com/telus/tds-community/commit/5d3c382))
* **community-pagination:** refactoring code ([e7394ee](https://github.com/telus/tds-community/commit/e7394ee))
* **packages:** adding hash to pagination deps ([1982942](https://github.com/telus/tds-community/commit/1982942))
