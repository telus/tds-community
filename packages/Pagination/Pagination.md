```jsx
<Pagination>
  <Pagination.Panel>Content 1</Pagination.Panel>
  <Pagination.Panel>Content 2</Pagination.Panel>
  <Pagination.Panel>Content 3</Pagination.Panel>
  <Pagination.Panel>Content 4</Pagination.Panel>
  <Pagination.Panel>Content 5</Pagination.Panel>
  <Pagination.Panel>Content 6</Pagination.Panel>
  <Pagination.Panel>Content 7</Pagination.Panel>
  <Pagination.Panel>Content 8</Pagination.Panel>
  <Pagination.Panel>
    <div>
      <p>This is a paragraph on the 9th panel</p>
    </div>
  </Pagination.Panel>
</Pagination>
```
