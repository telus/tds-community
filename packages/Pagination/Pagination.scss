@import '~@tds/core-responsive/responsive';
@import '~@tds/core-colours/colours';

.container {
  opacity: 1;
}

.controls {
  clear: both;
  display: flex;
  margin-top: 20px;
  position: relative;
  justify-content: center;
  width: 100%;
}

.next, .nextHidden, .previous, .previousHidden {
  color: $color-telus-purple;
  padding: 0 8px;
  font-size: 1rem;
  line-height: 16px;
  @include mq($from: md) {
    line-height: 22px;
  }
  button {
    color: $color-telus-purple;
    background-color: $color-white-lilac;
    border: none;
    width: 32px;
    height: 32px;
    text-align: center;
    cursor: pointer;
    @include mq($from: md) {
      background: none;
      width: auto;
      height: auto;
    }
    span.buttonLabel {
      display: none;
      @include mq($from: md) {
        display: inline-block;
      }
    }
  }
}

.nextHidden, .previousHidden {
  visibility: hidden;
}

.pagination {
  display: flex;
  li {
    display: inline-block;
    color: $color-telus-purple;
    background-color: $color-white-lilac;
    height: 32px;
    line-height: 29px;
    width: auto;
    min-width: 32px;
    @include mq($from: md) {
      border: 1px solid $color-white-lilac;
      min-width: 22px;
      height: 24px;
      line-height: 24px;
    }
    list-style: none;
    margin-right: 4px;
    text-align: center;

    button {
      background: none;
      border: none;
      color: $color-telus-purple;
      cursor: pointer;
      padding: 0 5px;
      min-width: 32px;
      width: auto;
      height: 32px;
      @include mq($from: md) {
        min-width: 22px;
        height: 24px;
      }
    }
    @include mq($from: md) {
      &:hover {
        border: 1px solid $color-telus-purple;
      }
    }
  }
  li.ellipsis {
    background-color: inherit;
    line-height: 22px;
    @include mq($from: md) {
      line-height: 15px;
    }
    border: none;
    &:hover {
      border: none;
    }
  }
  li.current {
    color: $color-white;
    line-height: 31px;
    @include mq($from: md) {
      border: 1px solid $color-telus-purple;
      line-height: 23px;
      min-width: 24px;
    }
    padding: 0 5px;
    background-color: $color-telus-purple;
  }
}
