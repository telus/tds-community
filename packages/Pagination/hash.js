import stringHash from 'string-hash'

const JSONSafeStringify = input => {
  let cache = []
  const result = JSON.stringify(input, (key, value) => {
    if (typeof value !== 'object' || value === null) {
      return value
    }
    if (cache.includes(value)) {
      // Circular reference found, discard key
      return undefined
    }
    // Store value in our collection
    cache.push(value)
    return value
  })
  cache = null // Enable garbage collection
  return result
}

const hash = obj => {
  if (!obj) return 0
  return typeof obj !== 'string' ? stringHash(JSONSafeStringify(obj)) : stringHash(obj)
}

export default hash
