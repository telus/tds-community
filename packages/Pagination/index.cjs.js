require('./dist/index.css')
const Pagination = require('./dist/index.cjs')

module.exports = Pagination
