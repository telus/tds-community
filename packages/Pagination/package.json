{
  "name": "@tds/community-pagination",
  "version": "2.0.2",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-a11y-content": "^2.1.5",
    "@tds/core-colours": "^2.2.1",
    "@tds/core-decorative-icon": "^2.6.4",
    "@tds/core-responsive": "^1.3.3",
    "@tds/util-helpers": "^1.2.0",
    "prop-types": "^15.6.2",
    "string-hash": "^1.1.3"
  }
}
