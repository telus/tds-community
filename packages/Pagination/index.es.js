import Pagination from './dist/index.es'

export default Pagination
