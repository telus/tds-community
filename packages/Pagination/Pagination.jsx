import React, { Component } from 'react'
import PropTypes from 'prop-types'

import safeRest from '@tds/shared-safe-rest'
import DecorativeIcon from '@tds/core-decorative-icon'
import Panels from './Panels'
import Panel from './Panel/Panel'
import styles from './Pagination.scss'
import hash from './hash'

/**
 * @version ./package.json
 */
class Pagination extends Component {
  constructor(props) {
    super(props)
    this.state = {
      current: 1,
      panels: this.props.children.length,
      showPrevious: false,
      showNext: true,
    }
    this.mapNumeric = this.mapNumeric.bind(this)
    this.handleButtonState = this.handleButtonState.bind(this)
    this.handleClick = this.handleClick.bind(this)
  }

  componentDidMount() {
    this.handleButtonState()
  }

  mapNumeric = () => {
    return this.props.children.map((item, i) => {
      const index = i + 1
      let { current } = this.state
      const { panels } = this.state
      current = parseInt(current, 10) || 0
      if (current === index) {
        return (
          <li key={hash(`${i}-1`)} className={styles.current}>
            {index}
          </li>
        )
      }

      if (panels < 7 || index === 1 || index === panels) {
        return (
          <li key={hash(`${i}-2`)} className={styles.regular}>
            <button value={index} onClick={e => this.handleClick(e)}>
              {index}
            </button>
          </li>
        )
      }
      if (
        index === current ||
        index === current + 1 ||
        index === current - 1 ||
        (current < 3 && index < 5) ||
        (current > panels - 2 && index > panels - 4)
      ) {
        return (
          <li key={hash(`${i}-3`)} className={styles.regular}>
            <button
              value={index}
              onClick={e => this.handleClick(e)}
              aria-label={`Go to panel number ${index}`}
            >
              {index}
            </button>
          </li>
        )
      }
      if (
        index === current - 2 ||
        index === current + 2 ||
        (current < 3 && index === 5) ||
        (current > panels - 2 && index === panels - 5)
      ) {
        return (
          <li key={hash(`${i}-4`)} className={styles.ellipsis}>
            ...
          </li>
        )
      }

      return null
    })
  }

  handleClick = e => {
    e.preventDefault()
    const value = parseInt(e.currentTarget.value, 10) || 0
    if (value > this.state.panels || value < 1) {
      return null
    }
    return this.setState({ current: value }, this.handleButtonState)
  }

  handleButtonState = () => {
    if (this.state.current !== 1) {
      this.setState({ showPrevious: true })
    } else {
      this.setState({ showPrevious: false })
    }
    if (this.state.current !== this.state.panels) {
      this.setState({ showNext: true })
    } else {
      this.setState({ showNext: false })
    }
  }

  render() {
    const { children, language, ...rest } = this.props
    const { current } = this.state
    const increaseNumber = parseInt(current + 1, 10)
    const decreaseNumber = parseInt(current - 1, 10)
    const PreviousText = language !== 'french' ? 'Previous' : 'Précédent'
    const NextText = language !== 'french' ? 'Next' : 'Suivant'
    const PreviousLabel =
      language !== 'french' ? 'Go to previous panel' : 'Aller au panneau précédent'
    const NextLabel = language !== 'french' ? 'Go to next panel' : 'Aller au prochain panneau'
    return (
      <div {...safeRest(rest)} className={styles.container}>
        <Panels {...rest}>{children[current - 1] && children[current - 1]}</Panels>
        <div className={styles.controls}>
          {this.state.showPrevious && (
            <p className={styles.previous}>
              <button
                value={decreaseNumber}
                onClick={e => this.handleClick(e)}
                aria-label={PreviousLabel}
              >
                <DecorativeIcon symbol="leftChevron" size={16} />{' '}
                <span className={styles.buttonLabel}>{PreviousText}</span>
              </button>
            </p>
          )}
          <ul className={styles.pagination}>{this.mapNumeric()}</ul>
          {this.state.showNext && (
            <p className={styles.next}>
              <button
                value={increaseNumber}
                onClick={e => this.handleClick(e)}
                aria-label={NextLabel}
              >
                <span className={styles.buttonLabel}>{NextText}</span>{' '}
                <DecorativeIcon symbol="chevron" size={16} />
              </button>
            </p>
          )}
        </div>
      </div>
    )
  }
}

Pagination.propTypes = {
  children: PropTypes.node.isRequired,
  language: PropTypes.string,
}

Pagination.defaultProps = {
  language: 'english',
}

Pagination.Panel = Panel

export default Pagination
