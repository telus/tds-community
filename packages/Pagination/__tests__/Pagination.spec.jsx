import React from 'react'
import { mount } from 'enzyme'

import Pagination from '../Pagination'

describe('Pagination', () => {
  const children = (
    <Pagination copy="en">
      <Pagination.Panel>Content 1</Pagination.Panel>
      <Pagination.Panel>Content 2</Pagination.Panel>
      <Pagination.Panel>Content 3</Pagination.Panel>
      <Pagination.Panel>Content 4</Pagination.Panel>
      <Pagination.Panel>Content 5</Pagination.Panel>
      <Pagination.Panel>Content 6</Pagination.Panel>
      <Pagination.Panel>Content 7</Pagination.Panel>
      <Pagination.Panel>Content 8</Pagination.Panel>
      <Pagination.Panel>
        <div>
          <p>This is a paragraph on the 9th panel</p>
        </div>
      </Pagination.Panel>
    </Pagination>
  )
  const doMount = () => mount(children)

  it('renders', () => {
    const pagination = doMount()

    expect(pagination).toMatchSnapshot()
  })

  it('does other things', () => {
    const pagination = doMount()

    expect(pagination).toExist()
  })

  it('does not allow custom CSS', () => {
    const pagination = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(pagination).not.toHaveProp('className', 'my-custom-class')
    expect(pagination).not.toHaveProp('style')
  })
})
