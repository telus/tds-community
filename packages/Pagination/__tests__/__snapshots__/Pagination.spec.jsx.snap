// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders 1`] = `
<Pagination
  language="English"
>
  <div
    className="container"
  >
    <Panel>
      <div>
        <Panel>
          <div>
            Content 1
          </div>
        </Panel>
      </div>
    </Panel>
    <div
      className="controls"
    >
      <p
        className="previousHidden"
      >
        <button
          aria-label="Go to previous panel"
          onClick={[Function]}
          value={0}
        >
          <DecorativeIcon
            size={16}
            symbol="leftChevron"
          >
            <Icon
              aria-hidden="true"
              size={16}
              symbol="leftChevron"
            >
              <i
                aria-hidden="true"
                className="TDS_Icon-modules__iconLeftChevron___37tW0 TDS_Icon-modules__icon___13xYd TDS_Icon-modules__size16___1pm12"
              />
            </Icon>
          </DecorativeIcon>
           
          <span
            className="buttonLabel"
          >
            Prev
          </span>
        </button>
      </p>
      <ul
        className="pagination"
      >
        <li
          className="current"
          key="193359721"
        >
          1
        </li>
        <li
          className="regular"
          key="193361898"
        >
          <button
            aria-label="Go to panel number 2"
            onClick={[Function]}
            value={2}
          >
            2
          </button>
        </li>
        <li
          className="regular"
          key="193361897"
        >
          <button
            aria-label="Go to panel number 3"
            onClick={[Function]}
            value={3}
          >
            3
          </button>
        </li>
        <li
          className="regular"
          key="193361896"
        >
          <button
            aria-label="Go to panel number 4"
            onClick={[Function]}
            value={4}
          >
            4
          </button>
        </li>
        <li
          className="ellipsis"
          key="193354536"
        >
          ...
        </li>
        <li
          className="regular"
          key="193361891"
        >
          <button
            aria-label="Go to panel number 9"
            onClick={[Function]}
            value={9}
          >
            9
          </button>
        </li>
      </ul>
      <p
        className="next"
      >
        <button
          aria-label="Go to next panel"
          onClick={[Function]}
          value={2}
        >
          <span
            className="buttonLabel"
          >
            Next
          </span>
           
          <DecorativeIcon
            size={16}
            symbol="chevron"
          >
            <Icon
              aria-hidden="true"
              size={16}
              symbol="chevron"
            >
              <i
                aria-hidden="true"
                className="TDS_Icon-modules__iconChevron___u32Fg TDS_Icon-modules__icon___13xYd TDS_Icon-modules__size16___1pm12"
              />
            </Icon>
          </DecorativeIcon>
        </button>
      </p>
    </div>
  </div>
</Pagination>
`;
